### DISCUSSION TITLE
如何才能像core项目那样，编写的模块让web和electron都可以引用？

### DISCUSSION THREAD

**Discussion by hanbingzi** [opensumi/core#4025] [Q&A]
我编写的模块，会同时打包成为web版本和electron版本，如果我使用ide-startup或ide-electron项目，都只能打包成为一种版本，我希望像core那样，编写的代码是一个模块，可以被ide-startup和ide-electron项目引用，这样我就能在一个地方写代码，多种方式打包，如何才能实现这种方式？或者能给一下大体建议吗？

**Comment by Ricbet** [comment]
为什么要将模块打包成两个版本呢？opensumi 模块本身并没有强区分 web 或 electron，有也只是在代码里根据 `isElectronRenderer` 字段去决定渲染方式。
core 是使用 lerna 进行包管理的，你也可以将你自己写的模块发布成一个 npm 包，然后在 web 和 electron 项目中引入

**Comment by erha19** [comment]
@hanbingzi 目前 Core 项目中实现跨平台复用的逻辑基本上为逻辑判断，即通过环境判断执行逻辑，通过 `if..else` 逻辑去让特定代码在对应平台执行。

常用的环境判断为 `isElectronRenderer` 和 `isWindows...` 等，额外的，在 WebIDE 中，有时候还需要判断服务端的环境信息，在 ApplicationService 中又有 `backenOS` 等属性可以使用，可以参考一下。

**Reply by hanbingzi** [reply]
可能我说的意思没表达清楚，我现在在ide-startup下的modules编写代码，编写好之后，打包成为了一个web版本，现在同时要打包一个electron版本，我目前的做法是将我在ide-startup下modules写的代码复制到ide-electron的modules中，我希望像core这样，一个地方写，两个框架都能引用，我尝试用core的项目架构，把packages里面的代码删除了，放我自己的modules代码，尝试了几天没解决，所以来问下
